autopilot.go

```diff
@@ -132,26 +157,20 @@ type Autopilot struct {
 	// brought up.
 	startTime time.Time
 
-	// running is a simple bool to indicate whether the go routines to actually
-	// execute autopilot are currently running
-	running bool
-
 	// removeDeadCh is used to trigger the running autopilot go routines to
 	// find and remove any dead/failed servers
 	removeDeadCh chan struct{}
 
 	// reconcileCh is used to trigger an immediate round of reconciliation.
 	reconcileCh chan struct{}
 
-	// shutdown is a function that can be execute to shutdown a running
-	// autopilot's go routines.
-	shutdown context.CancelFunc
-	// done is a chan that will be closed when the running autopilot go
-	// routines have exited. Technically closing it is the very last
-	// thing done in the go routine but at that point enough state has
-	// been cleaned up that we would then allow it to be started
-	// immediately afterward
-	done chan struct{}
+	// execution is the information about the current autopilot execution
+	// Start will initialize this with the most recent execution
+	execution *execInfo
+
+	// execMutex implements a cancellable mutex that will be used to ensure
+	// that only one autopilot go routine is executing at once.
+	execMutex *mutex
```

Comment:
In terms of naming, I at first thought that `execMutex` was used to control access to `execution`, but eventually figured out that the `runLock` controls access to `execution` instead, thanks to one of your comments below. Maybe changing the `exec` prefix on `execMutex` would help to clarify that the mutex does not control access to `execution`.  Or `execution *execInfo` could change to `runInfo *runInfo` perhaps.

Reply:
I renamed a bunch of things and added hopefully better doc comments. `execMutex` became `leaderLock` amongst other things. You can see just the second commit of the PR if you just want to view the renaming (and I added an `IsRunning` method which will be necessary to make the integration test in consul easier)